File: src/lib/software/qbittorrent.py
from json import dumps
from re import compile, RegexFlag
from src.lib.software import BaseSoftware, SoftwareCategory
from PySide6.QtNetwork import QNetworkReply, QNetworkRequest
from PySide6.QtCore import Slot, QUrl, QByteArray, QJsonDocument

class QBitTorrent(BaseSoftware):
    def __init__(self):
        super().__init__()

        self.key = 'qbittorrent'
        self.name = 'qBittorrent'
        self.category = SoftwareCategory.Utility
        self.download_name = 'qbittorrent_x64_setup.exe'
        self.is_archive = False
        self.should_cache_url = False
        self.requires_admin = False
        self.icon = 'qbittorrent.png'
        self.homepage = 'https://qbittorrent.org'

        self._initial_url = QUrl('https://www.fosshub.com/qBittorrent.html')

    @Slot(QNetworkReply)
    def on_manager_finished(self, reply: QNetworkReply):
        reply.deleteLater()
        error = reply.error()
        if error != QNetworkReply.NetworkError.NoError:
            self.url_resolve_error.emit(self.ResolveError.URLResolveError)
            return

        text = reply.readAll()

        if reply.url() == self._initial_url:
            settings_pattern = compile(r'<script>\s+var settings =(.*)\s+</script>', RegexFlag.MULTILINE)
            match = settings_pattern.search(text.toStdString())
            if not match:
                self.url_resolve_error.emit(self.ResolveError.URLResolveError)
            else:
                json = QJsonDocument.fromJson(match.group(1).encode()).object()
                project_id = json['projectId']
                project_uri = json['pool']['u']
                project_files = json['pool']['f']

                latest_release = project_files[0]
                latest_file = latest_release['n']
                latest_id = latest_release['r']

                payload_json = dumps({
                    'projectId': project_id,
                    'releaseId': latest_id,
                    'projectUri': project_uri,
                    'fileName': latest_file,
                    'source': 'CF',
                })

                req = QNetworkRequest('https://api.fosshub.com/download/')
                req.setHeader(QNetworkRequest.KnownHeaders.ContentTypeHeader, 'application/json')

                self.manager.post(req, QByteArray(payload_json.encode()))
        else:
            json = QJsonDocument.fromJson(text).object()
            self.url_resolved.emit(json['data']['url'])

    def resolve_download_url(self):
        self.manager.get(QNetworkRequest(self._initial_url))

File: packages/app/src/services/software/definitions/qbittorrent.ts
import { ok, err } from 'neverthrow';
import { SoftwareCategory, DownloadUrlResolveError } from 'shared';
import type { ISoftwareDefinition } from 'shared';

export default class implements ISoftwareDefinition {
	public key = 'qbittorrent';
	public name = 'qBittorrent';
	public category = [SoftwareCategory.FileManagement, SoftwareCategory.Network, SoftwareCategory.Utility];
	public downloadName = 'qbittorrent_x64_setup.exe';
	public isArchive = false;
	public shouldCacheUrl = false;
	public requiresAdmin = false;
	public deprecated = false;
	public icon = 'qbittorrent.png';
	public homepage = 'https://qbittorrent.org';

	public async resolveDownloadUrl() {
		const payload = await this.getDownloadUrlPayload();
		if (!payload) {
			return err(DownloadUrlResolveError.Generic);
		}

		const res = await fetch('https://api.fosshub.com/download/', {
			method: 'POST',
			headers: {
				'content-type': 'application/json'
			},
			body: payload
		});
		if (!res.ok) {
			return err(DownloadUrlResolveError.HTTPResponseError);
		}

		const json = await res.json() as { data: { url: string; }; };

		return ok(json.data.url);
	}

	private async getDownloadUrlPayload() {
		const res = await fetch('https://www.fosshub.com/qBittorrent.html');
		if (!res.ok) {
			return null;
		}

		const text  = await res.text();
		const match = text.match(/<script>\s+var\s+settings\s+=(.*)\s+<\/script>/m);
		if (!match) {
			return null;
		}

		const json         = JSON.parse(match[1]);
		const projectId    = json.projectId as string;
		const projectUri   = json.pool.u as string;
		const projectFiles = json.pool.f as Array<{ n: string; r: string; }>;
		const latestFile   = projectFiles[0];

		return JSON.stringify({
			projectId,
			releaseId: latestFile.r,
			projectUri,
			fileName: latestFile.n,
			source: 'CF'
		});
	}
}
